Petition to keep `ti.pyfunc`

I understand that the plan is to remove `ti.pyfunc` in 1.6.0, but I found this to be a very important feature which I use all the time, so I'd like to make a case for keeping it in taichi.

In #4414, @lin-hitonami explained that even without `ti.pyfunc`, you can still create functions that work in both taichi scope and python scope using the following pattern:
```py
def fmod_python(a, b):
    return a-ti.floor(a/b)*b

fmod_taichi = ti.func(fmod_python)
```
While this does work for the simplest scenarios, there are many use cases that this pattern cannot support:

* Multiple functions that calls each other. For example:
```py
xs = ti.field(float, shape=100)

@ti.pyfunc
def f(i):
    return xs[i] + 1

@ti.pyfunc
def g(i):
    return f(i) + xs[i] + 1

@ti.kernel
def k():
    print(g(3))

k()
```
The solution proposed in #4414 does not work here:
```py
xs = ti.field(float, shape=100)

def f(i):
    return xs[i] + 1

def g(i):
    return f(i) + xs[i] + 1

@ti.kernel
def k():
    print(ti.func(g)(3))

k()
```
because in `ti.func(g)`, the `f` function that it calls is not transformed by `ti.func`, so this triggers an error:
```
  File "c:\users\ldfra\code\taichi\taichi\python\taichi\lang\util.py", line 308, in wrapped
    assert in_python_scope(), \
AssertionError: __getitem__ cannot be called in Taichi-scope
```

* member functions of `@ti.data_oriented` classes:
Consider for example:
```py
@ti.data_oriented
class C:
    @ti.pyfunc
    def f(self, i):
        return i + 1

c = C()

print(c.f(1))

@ti.kernel
def k():
    print(c.f(1))

k()
```
this would be very awkward without `ti.pyfunc`. The only solution I see is by doing
```py
@ti.data_oriented
class C:
    def f(self, i):
        return i + 1

@ti.kernel
def k():
    print(ti.func(C.f)(c, 1))
```
which is quite bad in terms of readability

* member functions struct types (`@ti.dataclass`). This is the same as with `ti.data_oriented` classes, because members of struct types are also callable from both taichi and python scope.
 
These are all very common use cases, and not having `ti.pyfunc` really makes writing large-scale taichi programs a lot more awkward. So please consider keeping `ti.pyfunc`. Alternatively, I would also greatly appreciate if you could help me identify other patterns that could solve all of the three aforementioned problems!

cc @lin-hitonami @strongoier @k-ye @ailzhang @yuanming-hu

--- a/python/taichi/__init__.py
+++ b/python/taichi/__init__.py
@@ -52,7 +52,6 @@ __customized_deprecations__ = {
                   'lang.misc._serialize'),
     'block_dim': ('Please use ti.loop_config(block_dim=...) instead.',
                   'lang.misc._block_dim'),
-    'pyfunc': ('Please avoid using it.', 'lang.kernel_impl.pyfunc'),
     'TriMesh':
     ('Please import meshtaichi_patcher as Patcher and use Patcher.load_mesh(...) instead.',
      'lang.mesh._TriMesh'),

--- a/python/taichi/lang/kernel_impl.py
+++ b/python/taichi/lang/kernel_impl.py
@@ -1149,4 +1149,4 @@ def data_oriented(cls):
     return cls
 
 
-__all__ = ["data_oriented", "func", "kernel"]
+__all__ = ["data_oriented", "func", "kernel", "pyfunc"]

--- a/tests/python/test_api.py
+++ b/tests/python/test_api.py
@@ -76,13 +76,14 @@ user_api[ti] = [
     'j', 'jk', 'jkl', 'jl', 'k', 'kernel', 'kl', 'l', 'lang', 'length',
     'linalg', 'log', 'loop_config', 'math', 'max', 'mesh_local',
     'mesh_patch_idx', 'metal', 'min', 'ndarray', 'ndrange', 'no_activate',
-    'one', 'opengl', 'polar_decompose', 'pow', 'profiler', 'randn', 'random',
-    'raw_div', 'raw_mod', 'ref', 'rescale_index', 'reset', 'rgb_to_hex',
-    'root', 'round', 'rsqrt', 'select', 'set_logging_level', 'simt', 'sin',
-    'solve', 'sparse_matrix_builder', 'sqrt', 'static', 'static_assert',
-    'static_print', 'stop_grad', 'svd', 'sym_eig', 'sync', 'tan', 'tanh',
-    'template', 'tools', 'types', 'u16', 'u32', 'u64', 'u8', 'ui', 'uint16',
-    'uint32', 'uint64', 'uint8', 'vulkan', 'wasm', 'x64', 'x86_64', 'zero'
+    'one', 'opengl', 'polar_decompose', 'pow', 'profiler', 'pyfunc', 'randn',
+    'random', 'raw_div', 'raw_mod', 'ref', 'rescale_index', 'reset',
+    'rgb_to_hex', 'root', 'round', 'rsqrt', 'select', 'set_logging_level',
+    'simt', 'sin', 'solve', 'sparse_matrix_builder', 'sqrt', 'static',
+    'static_assert', 'static_print', 'stop_grad', 'svd', 'sym_eig', 'sync',
+    'tan', 'tanh', 'template', 'tools', 'types', 'u16', 'u32', 'u64', 'u8',
+    'ui', 'uint16', 'uint32', 'uint64', 'uint8', 'vulkan', 'wasm', 'x64',
+    'x86_64', 'zero'
 ]
 user_api[ti.ad] = [
     'FwdMode', 'Tape', 'clear_all_gradients', 'grad_for', 'grad_replaced',

